gr-iio/lib/device_source_impl.cc

```diff
@@ -0,0 +1,487 @@
+}
+
+void device_source_impl::channel_read(const struct iio_channel* chn,
+                                      void* dst,
+                                      size_t len)
+{
+    uintptr_t src_ptr, dst_ptr = (uintptr_t)dst, end = dst_ptr + len;
+    unsigned int length = iio_channel_get_data_format(chn)->length / 8;
+    uintptr_t buf_end = (uintptr_t)iio_buffer_end(buf);
+    ptrdiff_t buf_step = iio_buffer_step(buf) * (decimation + 1);
+
+    for (src_ptr = (uintptr_t)iio_buffer_first(buf, chn) + byte_offset;
+         src_ptr < buf_end && dst_ptr + length <= end;
+         src_ptr += buf_step, dst_ptr += length)
+        iio_channel_convert(chn, (void*)dst_ptr, (const void*)src_ptr);
+}
+
+void device_source_impl::refill_thread()
+{
+    boost::unique_lock<boost::mutex> lock(iio_mutex);
+    ssize_t ret;
+
+    for (;;) {
+        while (!please_refill_buffer)
+            iio_cond.wait(lock);
+
+        please_refill_buffer = false;
+
+        lock.unlock();
+        ret = iio_buffer_refill(buf);
+        lock.lock();
+
+        if (ret < 0)
+            break;
+
+        items_in_buffer = (unsigned long)ret / iio_buffer_step(buf);
+        byte_offset = 0;
+
+        iio_cond2.notify_all();
+    }
+
+    /* -EBADF happens when the buffer is cancelled */
+    if (ret != -EBADF) {
+
+        char buf[256];
+        iio_strerror(-ret, buf, sizeof(buf));
+        std::string error(buf);
+
+        std::cerr << "Unable to refill buffer: " << error << std::endl;
+    }
+
+    thread_stopped = true;
+    iio_cond2.notify_all();
+}
+
+int device_source_impl::work(int noutput_items,
+                             gr_vector_const_void_star& input_items,
+                             gr_vector_void_star& output_items)
+{
+    boost::unique_lock<boost::mutex> lock(iio_mutex);
+
+    if (thread_stopped)
+        return -1; /* EOF */
+
+    /* No items in buffer -> ask for a refill */
+    if (!please_refill_buffer && !items_in_buffer) {
+        please_refill_buffer = true;
+        iio_cond.notify_all();
+    }
+
+    while (please_refill_buffer) {
+        bool fast_enough =
+            iio_cond2.timed_wait(lock, boost::posix_time::milliseconds(timeout));
+        if (thread_stopped)
+            return -1; /* EOF */
+
+        if (!fast_enough) {
+            message_port_pub(port_id, pmt::mp("timeout"));
+
+            /* GNU Radio won't call our block anytime soon if the
+             * work() function returns 0; it can take up to 500ms
+             * before it gets called again. To avoid that, we make
+             * this block send itself a dummy message on its input
+             * system message port, so that the scheduler calls us
+             * again promptly. */
+#ifdef GR_VERSION_3_7_OR_LESS
+            pmt::pmt_t payload = pmt::from_long(0);
+#else
+            pmt::pmt_t payload = pmt::from_bool(false);
+#endif
+            pmt::pmt_t msg = pmt::cons(pmt::mp("done"), payload);
+            post(pmt::mp("system"), msg);
```

Comment:
Could you explain a bit about the interaction of the block's thread and the IO thread, and this hack, which seems to force the scheduler into busy-waiting. At the moment the scheduler polls with an interval of 100ms even if nothing changes from the perspective of the scheduler. This can lead to higher delays as mentioned in the comment, given the random scheduling of the OS. I think in the long run, the scheduler shouldn't poll at all. If there is new data available the block should be notified.
In this case, I'd expect the IO thread to notify the scheduler that there is new data available. This can be done through a message to the system port. Since this is thread-safe, this can also happen from the IO thread. Is there a reason why it's done like that?

Reply:
The threading model here provides backpressure to hardware and also tries to get data into the block buffers as soon as space is made available. As your comment and the code comment suggested this can create throughput slowdowns and latency issues due to the scheduler's update rate. So this hack forces the timeout control to be within the block itself and not from the scheduler.

We could move the message post call (hack) to the refill_thread which would break the polling effect from the work thread. Ideally, this block would completely ignore the scheduler and simply be paced by backpressure and hardware rates.

Reply:
Sorry, but I still don't get it. To me it looks like:
- it blocks in `work` (which I think is bad)
- if it times out after blocking, it just schedules again, i.e., busy-waiting (which I think is bad)
- I don't get the relevance of the "timeout" message. What does it indicate to downstream blocks. I thought is indicates discontinuities (lost samples), but that doesn't seem to be the case. It looks like it just means that the output buffer couldn't be filled in 100ms. Why is this relevant?

Why doesn't the IO thread fill the buffer continuously (and not only on demand) and `work` copies it to the output_buffer? If work copied all samples and returned 0, the IO thread should also send  a message when new samples are available so that the block is scheduled again.

Reply:
I'm not totally sure why blocking the work thread is bad. Does this stop the scheduler from servicing other blocks? However, I do agree that busy waiting is not good. The sections of code relating to the "fast_enough" variable and the "timeout" message where I believe from earlier versions of libiio which did not support timeouts.

> Why doesn't the IO thread fill the buffer continuously (and not only on demand) and `work` copies it to the output_buffer?
This can only happen when there is space to put data. So the work function has to backpressure the IO thread somehow, which is accomplished by signaling it when to run.

I rewrote the refill and work threads here which should prevent work blocking and remove busy waiting: https://gist.github.com/tfcollins/7c0f6fc389c7e3ae6c21a865f322335d
Does that seem more reasonable? We can also require libiio versions which have timeout features which has been available since 2016.

Reply:
Blocking in "work" happens in some blocks, like the throttle.

For a source block, it's totally OK to block since the block can't generate data anyway & thus there's nothing more that it can do.

For the sink block, blocking indefinitely can be problematic since the scheduler needs to spin every now and then to deal with incoming samples.

Thus: A timed block & followup check is the way to go regardless of source or sink [... note that I haven't looked at your code yet to see if this is what you're doing].

Reply:
_I wish there were a better way to handle the situation "I have no data to produce" in the scheduler, so that it doesn't spin calling "work" & getting nothing done time and again (which can take 100% of a CPU core; hence why it's bad)._

Reply:
If blocking work doesn't matter I can remove the refill_thread completely as the libiio buffer API should be enough

Reply:
I still think blocking is bad :-)
- A source block might have filled nearly the whole output buffer and is called again with space for only one sample in the output buffer. Then the block sleeps for a minute until a lot of samples are available but it is only able to write one. (Even though downstream might have consumed all  samples meanwhile). -> suboptimal
- It's not that you have to choose between busy-waiting or blocking. Why not just do it properly and only schedule the block when there are samples available ready to be written to the output  buffer?
- For now, there is one thread per block. But I thought we want to go forward with the runtime and decouple threads from blocks. Then, there will be one thread that handles multiple blocks. In that case, we just cannot block in work. I don't think that we should add more blocks that block in work, otherwise we have to rewrite whole GNU Radio before we can move forward.

Reply:
@bastibl up to now, the dogma has been that it's OK to block in work, especially in a source. In fact, all hardware drivers I can think of do this under the hood; there's also not that many alternatives. 

gr-uhd just calls `recv()`; that is a blocking call. How else should IIO do this? There's no way of telling the scheduler "come back in 20 ms" or so.

Reply:
I'm not sure if this is a general comment (how to do things with GNU Radio) or if you think this block is fine.

> @bastibl up to now, the dogma has been that it's OK to block in work

Looks like I didn't get the memo. Blocking in work caused many problems in the past (incoming messages are not recognized, shutdown didn't work, you are tight to the TPB scheduler, .....). We removed `delete_head_blocking` to stop people from doing it.
Please see this threads
https://github.com/gnuradio/gnuradio/pull/1029 
https://github.com/gnuradio/gnuradio/pull/797
https://lists.gnu.org/archive/html/discuss-gnuradio/2013-06/msg00546.html
https://lists.gnu.org/archive/html/discuss-gnuradio/2013-06/msg00552.html

> There's no way of telling the scheduler "come back in 20 ms" or so.

This block uses a dedicated IO thread to refill the buffer. It could just as well schedule the block (done=false message to system port) when data is available... Again, see https://github.com/gnuradio/gnuradio/pull/1029.

Anyway, I give up...